app/src/main/java/com/calculator/AdvancedCalcActivity.java

```diff
@@ -0,0 +1,510 @@
+            public void onClick(View view) {
+                values = "";
+                result = "";
+                label_values_adv.setText(values);
+                label_result_adv.setText(String.valueOf(result));
+            }
+        });
+
+        btn_point_adv.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                values = values + ".";
+                label_values_adv.setText(values);
+            }
+        });
+
+        btn_parenthesisOpen_adv.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                values = values + "(";
+                label_values_adv.setText(values);
+            }
+        });
+
+        btn_parenthesisClose_adv.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                values = values + ")";
+                label_values_adv.setText(values);
+            }
+        });
+
+        switch_mode_adv.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                if (!switch_mode_adv.isChecked()){
+                    Intent intent = new Intent();
+                    switch_mode_adv.setText(R.string.modeBasic);
+                    setResult(RESULT_OK, intent);
+                    finish();
+                }
+            }
+        });
+    }
+
+    private static String calc(String expression) {
+        expression = prepareExpression(expression);
+        int pos = 0;
+        //Extracting expression from braces, doing recursive call
+        //replace braced expression on result of it solving
+        if (-1 != (pos = expression.indexOf("("))) {
+
+            String subexp = extractExpressionFromBraces(expression, pos);
+            if (subexp.equals(syntaxError))
+
+                return subexp;
+
+            expression = expression.replace("(" + subexp + ")", calc(subexp));
+
+            return calc(expression);
+
+            //Three states for calculating sin cos exp
+            //input must be like sin0.7
+        } else if (-1 != (pos = expression.indexOf("sin"))) {
+            pos += 2;//shift index to last symbol of "sin" instead of first
+            String number = extractNumber(expression, pos, RIGHT_DIRECTION);
+            expression = expression.replace("sin" + number,
+                    Double.toString(Math.sin(Double.parseDouble(number))));
+
+            return calc(expression);
+
+        } else if (-1 != (pos = expression.indexOf("cos"))) {
+            pos += 2;
+            String number = extractNumber(expression, pos, RIGHT_DIRECTION);
+            expression = expression.replace("cos" + number,
+                    Double.toString(Math.cos(Double.parseDouble(number))));
+
+            return calc(expression);
+
+        } else if (-1 != (pos = expression.indexOf("exp"))) {
```

Comment:
If you have a long expression in a if block statement you could add these operations into a method for example:
```
if(isValidExpression("("))
....
   private boolean isValidExpression(String expression) {
      return -1 != (expression.indexOf(expression)
   }
```
and for 
**position = getPosition(expression);**
```
private int getPosition(String expression) {
  return expression.indexOf(expression);
}
```